tests/integration/test_time_manager.py

```diff
@@ -0,0 +1,191 @@
+tm1 = pp.TimeManager([0, 0.01, 0.02, 0.05], 0.01, constant_dt=True)
+tm2 = pp.TimeManager([0, pp.HOUR, 2 * pp.HOUR, 10 * pp.HOUR], pp.HOUR, constant_dt=True)
+
+
+@pytest.mark.parametrize("model_class", time_dependent_models)
+@pytest.mark.parametrize("time_manager", [tm0, tm1, tm2])
+def test_constant_dt(request, model_class, time_manager):
+    """Test models when constant time step is used"""
+
+    model = model_class(params={"use_ad": True, "time_manager": time_manager})
+    pp.run_time_dependent_model(model, model.params)
+    if "0" in request.node.callspec.id:
+        assert model.time_manager.is_constant
+        assert model.time_manager.time_init == 0
+        assert model.time_manager.time_final == 2
+        assert model.time_manager.dt == 1
+        assert model.time_manager.time == 2
+    elif "1" in request.node.callspec.id:
+        assert model.time_manager.is_constant
+        assert model.time_manager.time_init == 0
+        assert model.time_manager.time_final == 0.05
+        assert model.time_manager.dt == 0.01
+        assert model.time_manager.time == 0.05
+    elif "2" in request.node.callspec.id:
+        assert model.time_manager.is_constant
+        assert model.time_manager.time_init == 0
+        assert model.time_manager.time_final == 10 * pp.HOUR
+        assert model.time_manager.dt == pp.HOUR
+        assert model.time_manager.time == 10 * pp.HOUR
+
+
+# -----> Testing linear models
+@pytest.mark.parametrize("model_class", linear_time_dependent_models)
+def test_linear_non_constant_dt(model_class):
+    """Test linear models when non-constant time steps are used"""
+    time_manager = pp.TimeManager(schedule=[0, 1], dt_init=0.01)
+    model = model_class(params={"use_ad": True, "time_manager": time_manager})
+    with pytest.raises(NotImplementedError) as excinfo:
+        msg = "Time step cannot be adapted when the problem is linear."
+        pp.run_time_dependent_model(model, model.params)
+    assert msg in str(excinfo.value)
+
+
+# -----> Testing adaptation after convergence for non-linear models
+@pytest.mark.parametrize("model_class", nonlinear_time_dependent_models)
+@pytest.mark.parametrize("solution_type", ["parabolic", "trigonometric"])
+def test_time_step_adaptation_after_convergence(model_class, solution_type):
+    """Test iteration-based time step adaptation for non-linear models"""
+    time_manager = pp.TimeManager(
+        schedule=[0, 1],
+        dt_init=0.3,
+        dt_min_max=(0.1, 0.6),
+        iter_max=10,
+        iter_optimal_range=(4, 7),
+        iter_relax_factors=(0.7, 1.3),
+        recomp_factor=0.5,
+        recomp_max=5,
+        print_info=False,
+    )
+    params = {
+        "use_ad": True,
+        "time_manager": time_manager,
+        "solution_type": solution_type,
+        "num_cells": 5,
+    }
+    model = model_class(params=params)
+    pp.run_time_dependent_model(model, model.params)
+    np.testing.assert_equal(model.out["iterations"], [3, 3, 3])
```

Comment:
What is really being tested here? It seems like comparing to a known solution which could change depending on a whole range of components in the code without necessarily implying a bug.

Reply:
Good question. 

What I'm testing here is the behaviour of the time step control routine on a full non-linear model. In particular, here I'm testing that the solutions (for the given input parameters) should convergence after 3 iterations in each time step. By knowing this, I can also know the exact computed time steps by the algorithm.

To your point, the test might be sensible to changes. However, the model is posed on a Cartesian grid without fractures. Meaning, we should not expect really any changes. The exception to this claim might be changes in mpfa. But then again, we should not have major changes for Cartesian grids.

Reply:
@keileg Perhaps you can have a look at the test after Veljko's PR.